### ISSUE TITLE
noderedpi Node-RED[428]: ERROR: api.err.Invalid

### ISSUE THREAD

**Issue by lespatots** [Isaksson/node-red-contrib-unifi#3]
I have this installed and I believe I have setup the node correctly but I get the above error in the node-red log and a red flag that just says ERR. Is there any chance the error report could be improved so I know what is 'invalid'? Is there something in the conrtroller that needs to be set to allow the API?

**Comment by Isaksson** [comment]
Hello.
You do not need to make any changes to the controller.
What version of controller do you have?

Setup is easy.
Enter Server IP and Port
sitename 
username and password

and then you could select a command that you want to run.

Just inject something and you will receive the response.

**Comment by lespatots** [comment]
Hi,

THX for the fast reply.

The controller is 5.5.24 running on a local unRaid system as a Docker. I have the IP right and the port is 8443 (which I think was preset). I’m using the site name from the Settings page with the Username and Password from the same page. Whatever command it always just gives the error.

However, it occurs to me that using an ‘Inject’ node sets something in the message.payload (timestamp) but as message.payload.command isn’t set that can’t be it.

What might be good would be a check of the appropriate credentials when the node is initialised as happens for some other nodes such as MQTT or MYSQL?

Any other ideas on what I’ve done wrong, it does look simple as you say!!

Regards

Richard

**Comment by Isaksson** [comment]
I will update my test environment from 5.5.20 to 5.5.24 just to verify that version.
yes 8443 is the default value and also the site name default is default. You do not use the default name for your site?
Will soon post back when i have updated the controller.

**Comment by Isaksson** [comment]
It worked with version 5.5.24,
Please check the username and password, the error message api.err.Invalid indicates that the login failed.

**Comment by lespatots** [comment]
Hi,

Wow that was quick.

Oct 21 15:00:28 noderedpi Node-RED[21110]: ERROR: api.err.Invalid
Oct 21 15:00:32 noderedpi Node-RED[21110]: ERROR: api.err.Invalid

I cut and pasted the Username and Password from the Site Configuration and no the site isn’t ‘default’ and I cut and pasted that too!

It is a bit strange that there isn’t any more error logging. I have a ‘catch all’ node that sees nothing and the trailing ‘Debug’ node doesn’t output anything either!

**Comment by Isaksson** [comment]
The error message is for the underlying module that is making the requests to the Unifi controller, my Node is just a shell for Node-Red.
I just created a new user in the web GUI for Unifi with read permission and then just entered the same credentials in the Node-Red Node and it worked.

Maybe you could do the same? Just as a test, create a new user with a simple password.

**Comment by lespatots** [comment]
Hi,

We’re into unknown territory for me but I added another admin user and the error has changed 
noderedpi Node-RED[21110]: ERROR: api.err.NoSiteContext

Is that a clue?

**Comment by lespatots** [comment]
Hi,

We’re into unknown territory for me but I added another admin user and the error has changed 
noderedpi Node-RED[21110]: ERROR: api.err.NoSiteContext

Is that a clue?

**Comment by Isaksson** [comment]
That error message indicates that the sitename is incorrect.

**Comment by lespatots** [comment]
Again I cut and pasted after reverting to all lower case. Same error. Have attached screenshots if they can get through!

**Comment by lespatots** [comment]
![unifi node](https://user-images.githubusercontent.com/11376860/31852849-a48002c8-b67f-11e7-8a26-d884b3dfe7ab.jpeg)
![unifi site](https://user-images.githubusercontent.com/11376860/31852854-af3f33c8-b67f-11e7-81d8-62169ffba9dc.jpeg)

**Comment by Isaksson** [comment]
And the username RichardJ is working when you sign in to the controller software? (web gui)
The username is case sensitive.

**Comment by lespatots** [comment]
Yes, can log in no problem!

**Comment by lespatots** [comment]
Thanks for all your help. I will keep on trying to find the cause. Perhaps I’ll create a sandbox to play in.

I know it is a bit rude to make suggestions after all this but I’m from Yorkshire so that’s my excuse.

1) Elsewhere its a green dot for ‘ok’ and a red rectangle for an error
2) It is really hard to find the node red error log so any addition to the error text by the node would be really helpful
3) I’m sure some of the data could be checked in the node on the initialisation phase?
4) I wonder if you had considered using topic as the place for a command string?
5) An SQL node I have manages to throw an error message to ‘Debug’, that would be really useful

I appreciate that you are wrapping some underlying code, perhaps if you included a pointer to it in the READ.ME then it might be useful?

**Comment by Isaksson** [comment]
1: and elsewhere is a couple of other nodes that you use? I have seen many different states for this.
2: the error is written to console.log, but maybe its better to also send it out.
3: this node does connect on command and then disconnects, it does not hold a current session or something like that, I don't see any good reason to verify before the first command?
4: that is already possible, if you read the info tab of this node in Node-Red then you will find how to do it with a simple example. (msg.payload.command)
5: Yes I think that is a good suggestion

I have this in the readme file, Based on npm package node-unifi.

**Comment by lespatots** [comment]
Quick update.

The user password is not what is stored on the site page it is the login password! Probably obvious but not to a newbie (and I use 1Password so never type it in!). I am still stuck with the sitename and I wonder whether there isn’t some prefix or something else that is causing the problem. I notice that when I scan the ports, 8443 comes up as https, could that be an issue?

Thanks

**Comment by lespatots** [comment]
Hi,

A bit of googling and I cracked the other bit. The site doesn’t come from the site page it comes from the url used to access it. So lespatots won’t work but 0mwmi52a did!!!

How on earth does your implementation work?

**Comment by Isaksson** [comment]
I don't think you understand how the unifi controller works or how to access an API?
username and password is the username and password for a user in the Unifi controller, the site is the site that you have configured when you installed Unifi controller.

I dont think it is possible to make this any more simple then that.

**Comment by lespatots** [comment]
I think that is a bit harsh. For both the user and site identification I would have expected that this was the information taken from the Site page of the controller. I spelt out what I had entered early on. Nowhere in the short Node description does it say where this data is found. It seems I am not the only one who has foundered on the rather obscure site name gathered from the URL when trying to use the API.

I am sorry if I have wasted your time, the error messages are a bit obscure and limited and you didn’t figure out what I was doing wrong either!
